### PR TITLE
Add basic tar bubble functionality

### PR DESCRIPTION
**Pull request by vchadha**
This needs work, but figured I'd put up a PR while I figure out how to make this better T_T

Things need doin:
- Update tar bubble to match player size (so it encompasses it)
- Update button smash mechanic. It definitely does not work right.

### CODE REVIEW

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+using DG.Tweening;
+
+public class TarBubbleController : MonoBehaviour
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs]: We already talked about this, but this should probably be a managed controller

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        { 
+            // Move along y axis
+            currentPosition.y += deltaTime * speed;
+            
+            float degreesPerSecond = 360.0f / pathPeriod;
+            degrees = Mathf.Repeat(degrees + (deltaTime * degreesPerSecond), 360.0f);
+            float radians = degrees * Mathf.Deg2Rad;
+            
+            // Offset by cos wave
+            Vector3 offset = new Vector3(pathAmplitude * Mathf.Cos(radians), 0.0f, 0.0f);
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs]: Opt: Might be cool to add in some random variation to how big the amplitude is or maybe move it around?

But that may increase scope so take that with a grain of salt

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    void Start()
+    {
+        // Init values
+        currentPosition = transform.position;
+        pathAmplitude = Random.Range(pathAmplitudeMin, pathAmplitudeMax);
+        pathPeriod = Random.Range(pathPeriodMin, pathPeriodMax);
+        scaleDuration = Random.Range(scaleDurationMin, scaleDurationMax);
+
+        // DoTween scale to (1, 1) in scaleDuration seconds
+        gameObject.transform.DOScale(Vector3.one, scaleDuration);
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs]: Don't actually need to specify `gameObject` here! It's implicitly added if you just do `transform.DOScale`

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            // TODO: Resize bubble to size of player
+
+            // Stop movement
+            hasCollided = true;
+            currentPosition = collider.gameObject.transform.position;
+            transform.position = currentPosition;
+        }
+    }
+
+    public void Die()
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs:87]: "Die" 😂 that's so dramatic

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    private float pathAmplitude;
+    private float pathPeriod;
+    private float scaleDuration;
+
+    private Vector3 currentPosition;
+
+    private float degrees;
+
+    [SerializeField] private GameObject sourceTarTile;
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs]: Can just change this to type `TarTileController`, when you drag it in the editor it will automatically select that component reference!

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        // Reenable tar tile if we set to inactive
+        sourceTarTile.GetComponent<TarTileController>().SetActive(true);
+
+        // Destroy bubble
+        Destroy(gameObject);
+    }
+
+    public void SetActive(bool isActive)
+    {
+        this.isActive = isActive;
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs:103]: Why do you need this property?

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        currentPosition = transform.position;
+        pathAmplitude = Random.Range(pathAmplitudeMin, pathAmplitudeMax);
+        pathPeriod = Random.Range(pathPeriodMin, pathPeriodMax);
+        scaleDuration = Random.Range(scaleDurationMin, scaleDurationMax);
+
+        // DoTween scale to (1, 1) in scaleDuration seconds
+        gameObject.transform.DOScale(Vector3.one, scaleDuration);
+
+        // Call die function after ttl has expirded
+        Invoke("Die", timeToLive);
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs:45]: Pausing may cause troubles when it comes to using Invoke 👀

File: Assets/Scripts/GameElements/TarTileController.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    [SerializeField] private int spawnInterval;
+    [SerializeField] private bool isActive;
+
+    [SerializeField] private GameObject tarBubblePrefab;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        // Invoke spawn function every spawnInterval seconds
+        InvokeRepeating("SpawnBubble", spawnInterval, spawnInterval);
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarTileController.cs]: Should make this use ManagedController, but be careful with state changes

File: Assets/Scripts/GameElements/TarBubbleController.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        else
+        {
+            // currentPosition.y -= deltaTime * speed;
+            // transform.position = currentPosition;
+        }
+    }
+
+    protected void OnTriggerEnter2D(Collider2D collider)
+    {
+        if (collider.gameObject.CompareTag(Constants.playerTag) && isActive)
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/TarBubbleController.cs:72]: I really don't recommend having 2 handlers of the same OnTriggerEnter event. I'd recommend moving all of this logic into a function and calling that function from the OnTriggerEnter event of the player

File: Assets/Scripts/GameElements/PlayerController.cs [hunk 10]
```diff
@@ -160,6 +185,30 @@ private void CheckForWingInput()
...
+    private void CheckForPlayerMash()
+    {
+        float delta = (buttonPressedTime - lastButtonPressedTime);
+        if (delta == 0)
+        {
+            delta = 0.1f;
+        }
+
+        // TODO: one day make a better function for this jesus
+        int mashValue = (int) (1.0f / delta);
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/PlayerController.cs]: I mentioned this in the meeting but if we want it to just be a "count of how many times you've mashed the button" then you could just increment a counter every time the player clicks the button to flap their wings. But not sure if you want to go that route

**Review by justinro-underscore — COMMENTED** [review]
I know you're still in progress on this, but just leavin some comments here!

🍑

File: Assets/Scripts/GameElements/TarBubbleController.cs (hunk 2 above)
Comment (vchadha) [reply Assets/Scripts/GameElements/TarBubbleController.cs]: oooo like over time! that would be dopeee

File: Assets/Scripts/GameElements/TarBubbleController.cs (hunk 7 above)
Comment (vchadha) [reply Assets/Scripts/GameElements/TarBubbleController.cs:45]: ooooooooooooooooooooo good catch

Maybe I can do the on state change and set the bubble to inactive and then switch to active

File: Assets/Scripts/GameElements/TarBubbleController.cs (hunk 6 above)
Comment (vchadha) [reply Assets/Scripts/GameElements/TarBubbleController.cs:103]: I wanted it so the prefab bubble in the tar tile prefab wouldn't start moving around

BUT I don't like having it as a prefab in the prefab T_T
but when I tried to not have it that way I kept getting null values when trying to access a spawned prefab >:(

File: Assets/Scripts/GameElements/TarBubbleController.cs (hunk 9 above)
Comment (vchadha) [reply Assets/Scripts/GameElements/TarBubbleController.cs:72]: ooooo really?

File: Assets/Scripts/GameElements/PlayerController.cs (hunk 10 above)
Comment (vchadha) [reply Assets/Scripts/GameElements/PlayerController.cs]: yeah not a bad idea at all - better than this function lmao

File: Assets/Scripts/GameElements/TarBubbleController.cs (hunk 9 above)
Comment (justinro-underscore) [reply Assets/Scripts/GameElements/TarBubbleController.cs:72]: Yeah, just makes it hard to follow the logic

File: Assets/Scripts/GameElements/PlayerController.cs [hunk 11]
```diff
@@ -184,6 +203,27 @@ private void CheckForWingInput()
         }
     }
 
+    private void CheckForPlayerMash()
+    {
+        // Check if any key has been pressed and then any key released
+        // TODO: not great as if, improvement will be to do wing keys
+        // TODO: further improvement would be to detect flaps
+        // TODO: further further improvment is to rely on time - faster flaps = shorter trap time
+        if (buttonPressed != Input.anyKey)
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/PlayerController.cs]: You may want to do Input.anyKeyDown, since that would bypass this `buttonPressed` variable entirely!

File: Assets/Scripts/GameElements/PlayerController.cs [hunk 12]
```diff
@@ -413,4 +453,31 @@ private void SetInvulnerable()
...
+    }
+
+    public void TrapPlayer()
+    {
+        isPlayerTrapped = true;
+        
+        gripDrainScalar = gripDrainScalar * gripAmplifyingFactor;
+        innerWingForce = innerWingForce * wingDampeningFactor;
+        outerWingForce = outerWingForce * wingDampeningFactor;
+        lateralWingForce = lateralWingForce * wingDampeningFactor;
```
Comment (justinro-underscore) [inline Assets/Scripts/GameElements/PlayerController.cs]: I suppose this is fine, but I would prefer if `gripDrainScalar` and all the other forces were their own private variables rather than reassigning them. Then you could make the variables that we define in the editor something like `initGripDrainScalar`

File: Assets/Scripts/GameElements/TarBubbleController.cs (hunk 6 above)
Comment (justinro-underscore) [reply Assets/Scripts/GameElements/TarBubbleController.cs:103]: That's weird. Well if it works it works 🤷

File: Assets/Scripts/SceneControllers/PlayController.cs [hunk 13]
```diff
@@ -67,7 +67,13 @@ override protected void SceneUpdate()
         }
         if (Input.GetKeyDown(KeyCode.Backspace) && (State == PlayState.PAUSE || State == PlayState.RUNNING))
         {
+            // Toggle dotweens!
```
Comment (justinro-underscore) [inline Assets/Scripts/SceneControllers/PlayController.cs:70]: Thank you for adding 👍

**Review by justinro-underscore — APPROVED** [review]
The comments I have are pretty minor so you can decide what to implement and whatnot!

File: Assets/Scripts/GameElements/PlayerController.cs (hunk 12 above)
Comment (vchadha) [reply Assets/Scripts/GameElements/PlayerController.cs]: ahhh gotcha

**Review by justinro-underscore — APPROVED** [review]
Lovely 👏 🎬
